README.md
[![test](https://github.com/mauricioabreu/manifest-modifier/actions/workflows/CI.yml/badge.svg)](https://github.com/mauricioabreu/manifest-modifier/actions/workflows/CI.yml)
[![Crates.io](https://img.shields.io/crates/v/manifest-filter?label=manifest-filter)](https://crates.io/crates/manifest-filter)
[![Crates.io](https://img.shields.io/crates/v/manifest-server?label=manifest-server)](https://crates.io/crates/manifest-server)
[![manifest-filter docs](https://docs.rs/manifest-filter/badge.svg)](https://docs.rs/manifest-filter)

# Manifest modifier

*Manifest Modifier* is a work-in-progress project to modify video manifests.

Why? Video is a bit complex. Some manifests won't run on some devices because of the frame rate, or the bitrate, or other tags that may affect playback.

![manifest_modifier](/assets/manifest_modifier.png)

The image above is a perfect example that describes am usual problem: some devices can't play 60fps video. In order to solve this situation, we can use `manifest-modifier` to rewrite the manifest right before sending it to the users.

There are two ways to use this project, either as a lib or a server. This project is dividied into two crates: `manifest-filter` and `manifest-server`. `manifest-server` is a server built on top of [axum](https://github.com/tokio-rs/axum) and [m3u8-rs](https://github.com/rutgersc/m3u8-rs). It can be used without requiring advanced knowledge of the Rust programming language.

`manifest-filter` is the Rust code behind `manifest-server`. If you running your own server and can't use the `manifest-server`, no worries, you can use the same features by calling Rust code directly:

```rust
use manifest_filter::Master;
use std::io::Read;

let mut file = std::fs::File::open("manifests/master.m3u8").unwrap();
let mut content: Vec<u8> = Vec::new();
file.read_to_end(&mut content).unwrap();

let (_, master_playlist) = m3u8_rs::parse_master_playlist(&content).unwrap();
let mut master = Master {
    playlist: master_playlist,
};
master.filter_fps(Some(30.0));
 ```

The result should be  something like this

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio-aach-96",LANGUAGE="en",NAME="English",DEFAULT=YES,AUTOSELECT=YES,CHANNELS="2"
#EXT-X-STREAM-INF:BANDWIDTH=600000,AVERAGE-BANDWIDTH=600000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=384x216,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=249984.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=800000,AVERAGE-BANDWIDTH=800000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=768x432,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=1320960.m3u8
```

## Installation

The binary for `manifest-server` is `manifest_server`

You must have `cargo` to install the manifest server:

```
$ cargo install manifest-server
```

There is no other way to install the server right now.

## Usage

For the server to work, you need to export a variable `LISTEN_ADDRESS`

```
$ LISTEN_ADDRESS=127.0.0.1:3000 manifest_server
```

## Features

### Master playlist

**Bandwidth** - filter variants based on min and max values.

Request:

```
curl --request POST \
  --url 'http://localhost:3000/master?min_bitrate=800000&max_bitrate=2000000' \
  --header 'content-type: text/html; charset=UTF-8' \
  --header 'user-agent: vscode-restclient' \
  --data '< ../manifest-filter/manifests/master.m3u8'
```

Response:

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio-aach-96",LANGUAGE="en",NAME="English",DEFAULT=YES,AUTOSELECT=YES,CHANNELS="2"
#EXT-X-STREAM-INF:BANDWIDTH=800000,AVERAGE-BANDWIDTH=800000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=768x432,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=1320960.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1500000,AVERAGE-BANDWIDTH=1500000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=1280x720,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=3092992.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2000000,AVERAGE-BANDWIDTH=2000000,CODECS="mp4a.40.5,avc1.640029",RESOLUTION=1920x1080,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=4686976.m3u8
```

<details>
<summary>Original playlist</summary>

As you can see, the original playlist was slightly different:

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio-aach-96",LANGUAGE="en",NAME="English",DEFAULT=YES,AUTOSELECT=YES,CHANNELS="2"
#EXT-X-STREAM-INF:BANDWIDTH=600000,AVERAGE-BANDWIDTH=600000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=384x216,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=249984.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=800000,AVERAGE-BANDWIDTH=800000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=768x432,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=1320960.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1500000,AVERAGE-BANDWIDTH=1500000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=1280x720,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=3092992.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2000000,AVERAGE-BANDWIDTH=2000000,CODECS="mp4a.40.5,avc1.640029",RESOLUTION=1920x1080,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=4686976.m3u8
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=37000,CODECS="avc1.64001F",RESOLUTION=384x216,URI="keyframes/variant-video=249984.m3u8"
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=193000,CODECS="avc1.64001F",RESOLUTION=768x432,URI="keyframes/variant-video=1320960.m3u8"
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=296000,CODECS="avc1.64001F",RESOLUTION=1280x720,URI="keyframes/variant-video=2029952.m3u8"
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=684000,CODECS="avc1.640029",RESOLUTION=1920x1080,URI="keyframes/variant-video=4686976.m3u8"
```
</details>

**Frame rate** - filter variants based on a predefined *fps*:

Request:

```
curl --request POST \
  --url 'http://localhost:3000/master?rate=60' \
  --header 'content-type: text/html; charset=UTF-8' \
  --header 'user-agent: vscode-restclient' \
  --data '< ../manifest-filter/manifests/master.m3u8'
```

Response:

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio-aach-96",LANGUAGE="en",NAME="English",DEFAULT=YES,AUTOSELECT=YES,CHANNELS="2"
#EXT-X-STREAM-INF:BANDWIDTH=1500000,AVERAGE-BANDWIDTH=1500000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=1280x720,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=3092992.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2000000,AVERAGE-BANDWIDTH=2000000,CODECS="mp4a.40.5,avc1.640029",RESOLUTION=1920x1080,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=4686976.m3u8
```

<details>
<summary>Original playlist</summary>

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio-aach-96",LANGUAGE="en",NAME="English",DEFAULT=YES,AUTOSELECT=YES,CHANNELS="2"
#EXT-X-STREAM-INF:BANDWIDTH=600000,AVERAGE-BANDWIDTH=600000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=384x216,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=249984.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=800000,AVERAGE-BANDWIDTH=800000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=768x432,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=1320960.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=1500000,AVERAGE-BANDWIDTH=1500000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=1280x720,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=3092992.m3u8
#EXT-X-STREAM-INF:BANDWIDTH=2000000,AVERAGE-BANDWIDTH=2000000,CODECS="mp4a.40.5,avc1.640029",RESOLUTION=1920x1080,FRAME-RATE=60,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
variant-audio_1=96000-video=4686976.m3u8
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=37000,CODECS="avc1.64001F",RESOLUTION=384x216,URI="keyframes/variant-video=249984.m3u8"
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=193000,CODECS="avc1.64001F",RESOLUTION=768x432,URI="keyframes/variant-video=1320960.m3u8"
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=296000,CODECS="avc1.64001F",RESOLUTION=1280x720,URI="keyframes/variant-video=2029952.m3u8"
#EXT-X-I-FRAME-STREAM-INF:BANDWIDTH=684000,CODECS="avc1.640029",RESOLUTION=1920x1080,URI="keyframes/variant-video=4686976.m3u8"
```
</details>

### Media playlist

**DVR** - remove segments (backwards) based on the duration (seconds)

```
curl --request POST \
  --url 'http://localhost:3000/media?dvr=15' \
  --header 'content-type: text/html; charset=UTF-8' \
  --header 'user-agent: vscode-restclient' \
  --data '< ../manifest-filter/manifests/media.m3u8'
```

Response:

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-TARGETDURATION:8
#EXT-X-MEDIA-SEQUENCE:320035373
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035703.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035702.ts
#EXT-X-PROGRAM-DATE-TIME:2020-09-15T14:01:39.133333+00:00
#EXT-X-CUE-IN
#EXTINF:5.8666, no desc
variant-audio_1=96000-video=249984-320035701.ts
```

<details>
<summary>Original playlist</summary>

```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-MEDIA-SEQUENCE:320035356
#EXT-X-INDEPENDENT-SEGMENTS
#EXT-X-TARGETDURATION:8
#EXT-X-PROGRAM-DATE-TIME:2020-09-15T13:32:55Z
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035684.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035685.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035686.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035687.ts
#EXTINF:4.1333, no desc
variant-audio_1=96000-video=249984-320035688.ts
#EXT-X-DATERANGE:ID="4026531847",START-DATE="2020-09-15T14:00:39.133333Z",PLANNED-DURATION=60,SCTE35-OUT=0xFC3025000000000BB800FFF01405F00000077FEFFE0AF311F0FE005265C0000101010000817C918E
#EXT-X-CUE-OUT:60
#EXT-X-PROGRAM-DATE-TIME:2020-09-15T14:00:39.133333Z
#EXTINF:5.8666, no desc
variant-audio_1=96000-video=249984-320035689.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035690.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035691.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035692.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035693.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035694.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035695.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035696.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035697.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035698.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035699.ts
#EXTINF:4.1333, no desc
variant-audio_1=96000-video=249984-320035700.ts
#EXT-X-CUE-IN
#EXT-X-PROGRAM-DATE-TIME:2020-09-15T14:01:39.133333Z
#EXTINF:5.8666, no desc
variant-audio_1=96000-video=249984-320035701.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035702.ts
#EXTINF:5, no desc
variant-audio_1=96000-video=249984-320035703.ts
```
</details>

## Tests

```
cargo test
```

## Lint

```
cargo clippy
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [manifest-server-v0.1.2] - 2022-10-14
### Documentation

- How to install and use the server
- Document server and its params

### Features

- Handle shutdown signal

### Miscellaneous Tasks

- Change how diff is generated

## [manifest-server-v0.1.1] - 2022-10-13

### Bug Fixes

- Check for possible out of bounds when choosing initial variant

### Documentation

- Add crates.io badge
- Add docs badge
- Label badges for each project
- Add a changelog file

### Miscellaneous Tasks

- Upgrade manifest-filter dependency for manifest-server
- Fix Cargo.toml in order to make it installable
- Automatic changelog

### Refactor

- Reuse test functions to build master/media playlists
- Replace structs with Option values

### Styling

- Format code

<!-- generated by git-cliff -->


manifest-server/src/main.rs
//! manifest-server is an HTTP server you can use to modify video manifests.
//! It is built on top of axum and relies on manifest handlers provided by `manifest-filter`.
//!
//! There are two routes: `/master` and `/media`
//!
//! [Params] describe the available parameters you can use to modify video manifests.
//! Parameters are all optional and they are passed as query parameters. Example:
//!
//! ```not_rust
//! /master?min_bitrate=800000&max_bitrate=2000000
//! ```

use axum::{
    body::Bytes, extract::Query, http::StatusCode, response::IntoResponse, routing::post, Router,
};
use serde::Deserialize;
use std::env;
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();

    let app = Router::new()
        .route("/master", post(modify_master))
        .route("/media", post(modify_media));
    let addr = env::var("LISTEN_ADDRESS").expect("env var LISTEN_ADDRESS is not set");
    let socket_addr = addr
        .parse::<SocketAddr>()
        .expect("value for LISTEN_ADDRESS must be like 127.0.0.1:3000");
    tracing::debug!("listening on {}", socket_addr);
    axum::Server::bind(&socket_addr)
        .serve(app.into_make_service())
        .with_graceful_shutdown(shutdown_signal())
        .await
        .unwrap();
}

/// Tokio signal handler that will wait for a user to press CTRL+C
async fn shutdown_signal() {
    tokio::signal::ctrl_c()
        .await
        .expect("Expect shutdown signal handler");
    println!("signal shutdown");
}

async fn modify_master(params: Query<Params>, body: Bytes) -> impl IntoResponse {
    match manifest_filter::load_master(&body) {
        Ok(pl) => {
            let mut master = manifest_filter::Master { playlist: pl };
            master
                .filter_bandwidth(params.min_bitrate, params.max_bitrate)
                .filter_fps(params.rate)
                .first_variant_by_index(params.variant_index)
                .first_variant_by_closest_bandwidth(params.closest_bandwidth);

            let mut v: Vec<u8> = Vec::new();
            master.playlist.write_to(&mut v).unwrap();

            (StatusCode::OK, String::from_utf8(v).unwrap())
        }
        Err(e) => (StatusCode::BAD_REQUEST, e),
    }
}

async fn modify_media(params: Query<Params>, body: Bytes) -> impl IntoResponse {
    match manifest_filter::load_media(&body) {
        Ok(pl) => {
            let mut media = manifest_filter::Media { playlist: pl };
            media
                .filter_dvr(params.dvr)
                .trim(params.trim_start, params.trim_end);

            let mut v: Vec<u8> = Vec::new();
            media.playlist.write_to(&mut v).unwrap();

            (StatusCode::OK, String::from_utf8(v).unwrap())
        }
        Err(e) => (StatusCode::BAD_REQUEST, e),
    }
}

/// Query string params used to modify master and media manifests
#[derive(Debug, Deserialize, Default)]
struct Params {
    /// Min bitrate present in the master manifest
    min_bitrate: Option<u64>,
    /// Max bitrate present in the master manifest
    max_bitrate: Option<u64>,
    /// Frame rate allowed to be present in the master manifest
    rate: Option<f64>,
    /// DVR in seconds
    dvr: Option<u64>,
    /// Trim start
    trim_start: Option<u64>,
    /// Trim end
    trim_end: Option<u64>,
    /// Which index must be set as the first media manifest
    variant_index: Option<u64>,
    /// Which bandwidth must be set as the first media manifest
    closest_bandwidth: Option<u64>,
}


manifest-filter/src/lib.rs
//! manifest-filter is a lib used to modify video manifests.
//!
//! # Table of contents
//!
//! - [Features](#features)
//! - [Examples](#examples)
//!
//! # Features
//!
//! - Modify master playlists (filter variants by bandwidth, fps, etc)
//! - Modify media playlists (DVR, trim segments)
//!
//! More features are coming soon.
//!
//! # Examples
//!
//! You can try the example below, used to filter only the variants that are 30fps.
//!
//! ```rust
//! use manifest_filter::Master;
//! use std::io::Read;
//!
//! let mut file = std::fs::File::open("manifests/master.m3u8").unwrap();
//! let mut content: Vec<u8> = Vec::new();
//! file.read_to_end(&mut content).unwrap();
//!
//! let (_, master_playlist) = m3u8_rs::parse_master_playlist(&content).unwrap();
//! let mut master = Master {
//!     playlist: master_playlist,
//! };
//! master.filter_fps(Some(30.0));
//! ```
//!
//! The result should be something like this
//!
//! ```not_rust
//! #EXTM3U
//! #EXT-X-VERSION:4
//! #EXT-X-MEDIA:TYPE=AUDIO,GROUP-ID="audio-aach-96",LANGUAGE="en",NAME="English",DEFAULT=YES,AUTOSELECT=YES,CHANNELS="2"
//! #EXT-X-STREAM-INF:BANDWIDTH=600000,AVERAGE-BANDWIDTH=600000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=384x216,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
//! variant-audio_1=96000-video=249984.m3u8
//! #EXT-X-STREAM-INF:BANDWIDTH=800000,AVERAGE-BANDWIDTH=800000,CODECS="mp4a.40.5,avc1.64001F",RESOLUTION=768x432,FRAME-RATE=30,AUDIO="audio-aach-96",CLOSED-CAPTIONS=NONE
//! variant-audio_1=96000-video=1320960.m3u8
//! ```
//!
//! All functions can be chained. Call `filter_fps` to first remove variants with
//! a frame rate different of the one choosen and call `filter_bandwith` right after to
//! remove variants thare don't fit into the max/min range expected.
//! The sampe applies for `Media`.

use m3u8_rs::{MasterPlaylist, MediaPlaylist, Playlist};

pub fn load_master(content: &[u8]) -> Result<MasterPlaylist, String> {
    match m3u8_rs::parse_playlist(content) {
        Result::Ok((_, Playlist::MasterPlaylist(pl))) => Ok(pl),
        Result::Ok((_, Playlist::MediaPlaylist(_))) => Err("must be a master playlist".to_string()),
        Result::Err(e) => Err(e.to_string()),
    }
}

pub fn load_media(content: &[u8]) -> Result<MediaPlaylist, String> {
    match m3u8_rs::parse_playlist(content) {
        Result::Ok((_, Playlist::MediaPlaylist(pl))) => Ok(pl),
        Result::Ok((_, Playlist::MasterPlaylist(_))) => Err("must be a media playlist".to_string()),
        Result::Err(e) => Err(e.to_string()),
    }
}

/// `Master` holds a reference to the master playlist. All
/// functions implemented by this struct can be chained.
pub struct Master {
    pub playlist: MasterPlaylist,
}

/// `Media` holds a reference to the media playlist. All
/// functions implemented by this struct can be chained.
pub struct Media {
    pub playlist: MediaPlaylist,
}

impl Master {
    /// Filter variants from a master playlist based on the frame rate passed.
    pub fn filter_fps(&mut self, rate: Option<f64>) -> &mut Self {
        if let Some(r) = rate {
            self.playlist.variants.retain(|v| v.frame_rate == Some(r));
        }
        self
    }

    /// Filter variants from a master playlist based on the bandwidh passed.
    ///
    /// Variants can be filtered using `min` and `max` values for bandwidth.
    ///
    /// There's no need to pass a `min` value if you don't need to. The
    /// same happens for `max` value. For `min` we will set to zero by default
    /// and for the `max` we'll use the `u64::MAX` value.
    pub fn filter_bandwidth(&mut self, min: Option<u64>, max: Option<u64>) -> &mut Self {
        let min = min.unwrap_or(0);
        let max = max.unwrap_or(u64::MAX);

        self.playlist
            .variants
            .retain(|v| v.bandwidth >= min && v.bandwidth <= max);
        self
    }

    /// Set the first variant by index to appear in the playlist for the one that
    /// best suites the device needs. Most of the times such feature will
    /// be used to skip the initial variant (too low for some devices).
    ///
    /// If the `index` passed in could cause "out of bounds" error, the playlist
    /// will keep untouched.
    ///
    /// # Arguments
    /// * `index` - an Option containing the index you want to be the first variant. Variants will be swapped.
    pub fn first_variant_by_index(&mut self, index: Option<u64>) -> &mut Self {
        if let Some(i) = index {
            if i as usize <= self.playlist.variants.len() {
                self.playlist.variants.swap(0, i.try_into().unwrap());
            }
        }
        self
    }

    /// Set the first variant by closes bandwidth to appear in the playlist for the one that
    /// best suites the device needs. Most of the times such feature will
    /// be used to skip the initial variant (too low for some devices).
    ///
    /// # Arguments
    /// * `closest_bandwidth` - an Option containing an approximate bandwidth value you want for the first variant.
    pub fn first_variant_by_closest_bandwidth(
        &mut self,
        closest_bandwidth: Option<u64>,
    ) -> &mut Self {
        if let Some(c) = closest_bandwidth {
            let (idx, _) = self
                .playlist
                .variants
                .iter()
                .enumerate()
                .min_by_key(|(_, v)| (c as i64 - v.bandwidth as i64).abs())
                .unwrap();
            let fv = self.playlist.variants.remove(idx);
            self.playlist.variants.insert(0, fv);
        }
        self
    }
}

impl Media {
    /// Remove segments backwards from the media playlist, based on the duration
    /// set. The duration is in seconds.
    /// Media sequence will be affected: `<https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.3.2>`
    pub fn filter_dvr(&mut self, seconds: Option<u64>) -> &mut Self {
        let mut acc = 0;
        let total_segments = self.playlist.segments.len();

        if let Some(s) = seconds {
            self.playlist.segments = self
                .playlist
                .segments
                .iter()
                .rev()
                .take_while(|segment| {
                    acc += segment.duration as u64;
                    acc <= s
                })
                .cloned()
                .collect();
            self.playlist.media_sequence += (total_segments - self.playlist.segments.len()) as u64;
        }
        self
    }

    /// Remove segments from the media playlist, based on the start/end passed.
    /// Media sequence will be affected: `<https://datatracker.ietf.org/doc/html/rfc8216#section-4.3.3.2>`
    pub fn trim(&mut self, start: Option<u64>, end: Option<u64>) -> &mut Self {
        let start = start.unwrap_or(0);
        let end = end.unwrap_or_else(|| self.playlist.segments.len().try_into().unwrap());

        let segments = &self.playlist.segments[start as usize..end as usize];
        let total_segments = self.playlist.segments.len();
        self.playlist.segments = segments.to_vec();
        self.playlist.media_sequence += (total_segments - self.playlist.segments.len()) as u64;
        self
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Read;

    fn build_master() -> Master {
        let mut file = std::fs::File::open("manifests/master.m3u8").unwrap();
        let mut content: Vec<u8> = Vec::new();
        file.read_to_end(&mut content).unwrap();

        let (_, master_playlist) = m3u8_rs::parse_master_playlist(&content).unwrap();
        Master {
            playlist: master_playlist,
        }
    }

    fn build_media() -> Media {
        let mut file = std::fs::File::open("manifests/media.m3u8").unwrap();
        let mut content: Vec<u8> = Vec::new();
        file.read_to_end(&mut content).unwrap();

        let (_, media_playlist) = m3u8_rs::parse_media_playlist(&content).unwrap();
        Media {
            playlist: media_playlist,
        }
    }

    #[test]
    fn filter_60_fps() {
        let mut master = build_master();
        master.filter_fps(Some(60.0));

        assert_eq!(master.playlist.variants.len(), 2);
    }

    #[test]
    fn filter_min_bandwidth() {
        let mut master = build_master();

        master.filter_bandwidth(Some(800000), None);

        assert_eq!(master.playlist.variants.len(), 3);
    }

    #[test]
    fn filter_max_bandwidth() {
        let mut master = build_master();

        master.filter_bandwidth(None, Some(800000));

        assert_eq!(master.playlist.variants.len(), 6);
    }

    #[test]
    fn filter_min_and_max_bandwidth() {
        let mut master = build_master();

        master.filter_bandwidth(Some(800000), Some(2000000));

        assert_eq!(master.playlist.variants.len(), 3);
    }

    #[test]
    fn set_first_variant_by_index() {
        let mut master = build_master();

        master.first_variant_by_index(Some(1));

        assert_eq!(master.playlist.variants[0].bandwidth, 800000);
        assert_eq!(master.playlist.variants[1].bandwidth, 600000);
    }

    #[test]
    fn set_first_variant_by_out_of_bounds_index() {
        let mut master = build_master();

        master.first_variant_by_index(Some(100));

        assert_eq!(master.playlist.variants[0].bandwidth, 600000);
        assert_eq!(master.playlist.variants[1].bandwidth, 800000);
    }

    #[test]
    fn set_first_variant_by_closest_bandwidth() {
        let mut master = build_master();

        master.first_variant_by_closest_bandwidth(Some(1650000));
        assert_eq!(master.playlist.variants[0].bandwidth, 1500000);
        assert_eq!(master.playlist.variants[1].bandwidth, 600000);
    }

    #[test]
    fn filter_dvr_with_short_duration() {
        let mut media = build_media();

        media.filter_dvr(Some(15));

        assert_eq!(media.playlist.segments.len(), 3);
        assert_eq!(media.playlist.media_sequence, 320035373);
    }

    #[test]
    fn filter_dvr_with_long_duration() {
        let mut media = build_media();

        media.filter_dvr(Some(u64::MAX));

        assert_eq!(media.playlist.segments.len(), 20);
        assert_eq!(media.playlist.media_sequence, 320035356);
    }

    #[test]
    fn trim_media_playlist_with_start_only() {
        let mut media = build_media();

        media.trim(Some(5), None);

        assert_eq!(media.playlist.segments.len(), 15);
        assert_eq!(media.playlist.media_sequence, 320035361);
    }

    #[test]
    fn trim_media_playlist_with_end_only() {
        let mut media = build_media();

        media.trim(None, Some(5));

        assert_eq!(media.playlist.segments.len(), 5);
        assert_eq!(media.playlist.media_sequence, 320035371);
    }

    #[test]
    fn trim_media_playlist_with_start_and_end() {
        let mut media = build_media();

        media.trim(Some(5), Some(18));

        assert_eq!(media.playlist.segments.len(), 13);
        assert_eq!(media.playlist.media_sequence, 320035363);
    }
}
